libraries/c_sdk/standard/mqtt/src/iot_mqtt_operation.c

```diff
@@ -680,32 +682,53 @@ void _IotMqtt_ProcessKeepAlive( IotTaskPool_t pTaskPool,
-        bytesSent = pMqttConnection->pNetworkInterface->send( pMqttConnection->pNetworkConnection,
-                                                              pMqttConnection->pPingreqPacket,
-                                                              pMqttConnection->pingreqPacketSize );
+        /* Only send the PINGREQ if the keep-alive period has elapsed since the connection
+         * was last used. */
+        elapsedTime = IotClock_GetTimeMs() - pMqttConnection->lastMessageTime;
 
-        if( bytesSent != pMqttConnection->pingreqPacketSize )
+        if( elapsedTime < ( uint64_t ) pMqttConnection->keepAliveMs )
         {
-            IotLogError( "(MQTT connection %p) Failed to send PINGREQ.", pMqttConnection );
-            status = false;
+            IotLogDebug( "(MQTT connection %p) Connection was last used %llu ms ago, which "
+                         "is less than keep-alive period %lu ms. PINGREQ will not be sent.",
+                         pMqttConnection,
+                         ( unsigned long long ) elapsedTime,
+                         ( unsigned long ) pMqttConnection->keepAliveMs );
+
+            /* Schedule the next keep-alive job one keep-alive period after the last packet was sent. */
+            scheduleDelay = pMqttConnection->keepAliveMs - ( ( uint32_t ) elapsedTime );
         }
         else
         {
-            /* Assume the keep-alive will fail. The network receive callback will
-             * clear the failure flag upon receiving a PINGRESP. */
-            pMqttConnection->keepAliveFailure = true;
+            IotLogDebug( "(MQTT connection %p) Sending PINGREQ.", pMqttConnection );
 
-            /* Schedule a check for PINGRESP. */
-            pMqttConnection->nextKeepAliveMs = IOT_MQTT_RESPONSE_WAIT_MS;
+            /* Because PINGREQ may be used to keep the MQTT connection alive, it is
+             * more important than other operations. Bypass the queue of jobs for
+             * operations by directly sending the PINGREQ in this job. */
+            bytesSent = pMqttConnection->pNetworkInterface->send( pMqttConnection->pNetworkConnection,
+                                                                  pMqttConnection->pPingreqPacket,
+                                                                  pMqttConnection->pingreqPacketSize );
 
-            IotLogDebug( "(MQTT connection %p) PINGREQ sent. Scheduling check for PINGRESP in %d ms.",
-                         pMqttConnection,
-                         IOT_MQTT_RESPONSE_WAIT_MS );
+            if( bytesSent != pMqttConnection->pingreqPacketSize )
+            {
+                IotLogError( "(MQTT connection %p) Failed to send PINGREQ.", pMqttConnection );
+                status = false;
+            }
+            else
+            {
+                /* Update the timestamp of the last message on successful transmission. */
+                pMqttConnection->lastMessageTime = IotClock_GetTimeMs();
```

Comment:
`_IotMqtt_ProcessSend` uses `IotMutex_Lock( &( pMqttConnection->referencesMutex ) );` to protect access to `pMqttConnection->lastMessageTime`. Is it needed here too?

Reply:
+1,  mutex must be used because MQTTConnection is shared.

Reply:
Correct, however that mutex is already locked before this at line 680, and unlocked at the end of the function at line 846, so I had to remove the duplicate lock/unlock when copying this over. Same goes for the other access at line 687